src/input.rs

```diff
@@ -0,0 +1,497 @@
+    Key(KeyCode), // TODO: consider support for Key{scancode: u32}?
+    Mouse(MouseButton),
+}
+
+/// State of a button.
+#[derive(Debug, PartialEq, Eq, Hash, Copy, Clone)]
+pub enum State {
+    Pressed,
+    Hold,
+    Released,
+}
+
+/// Input Service
+pub struct Input {
+    mapper: Box<dyn std::any::Any + Send + Sync>,
+    states: HashMap<Button, State>,
+    mouse_scroll_delta: f32,
+    mouse_position: Vector2<f32>,
+    last_mouse_position: Vector2<f32>,
+    window_size: Vector2<f32>, // TODO: move to other struct or service
+}
+
+impl Input {
+    pub fn new(mapper: Box<dyn std::any::Any + Send + Sync>) -> Self {
+        Self {
+            mapper,
+            states: HashMap::new(),
+            mouse_scroll_delta: 0.0,
+            mouse_position: Vector2::new(0.0, 0.0),
+            last_mouse_position: Vector2::new(0.0, 0.0),
+            window_size: Vector2::new(0.0, 0.0),
+        }
+    }
+
+    /// Returns the status of the mapped action.
+    pub fn action_state<T>(&self, action: T) -> Option<State>
+    where
+        Self: ActionMapper<T>,
+        T: Copy + Eq + std::hash::Hash,
+    {
+        if let Some(button) = self.action_mapped(action) {
+            if let Some(state) = self.states.get(button) {
+                return Option::Some(*state);
+            }
+        }
+        Option::None
+    }
+
+
+    /// Checks if mapped action button is pressed.
+    pub fn action_pressed<T>(&self, action: T) -> bool
+    where
+        Self: ActionMapper<T>,
+        T: Copy + Eq + std::hash::Hash,
+    {
+        match self.action_state(action) {
+            Some(state) => state == State::Pressed,
+            None => false,
+        }
+    }
+
+    /// Checks if mapped action button is released.
+    pub fn action_released<T>(&self, action: T) -> bool
+    where
+        Self: ActionMapper<T>,
+        T: Copy + Eq + std::hash::Hash,
+    {
+        match self.action_state(action) {
+            Some(state) => state == State::Released,
+            None => false,
+        }
+    }
+
+    /// Checks if mapped button is pressed or hold.
+    pub fn action_hold<T>(&self, action: T) -> bool
+    where
+        Self: ActionMapper<T>,
+        T: Copy + Eq + std::hash::Hash,
+    {
+        match self.action_state(action) {
+            Some(state) => state == State::Hold || state == State::Pressed,
+            None => false,
+        }
+    }
+
+    /// Just a mapper getter.
+    pub fn get_mapper<T: 'static + Send + Sync>(&self) -> &T {
+        self.mapper.downcast_ref::<T>().unwrap()
+    }
+
+    /// Initialize input manager - it is requiered to get information about window size.
+    pub fn init(&mut self, window: &winit::window::Window) {
+        self.window_size = Vector2::new(window.inner_size().width as f32, window.inner_size().height as f32);
+    }
+
+    /// This method should be called in application.rs after render, so States from events will be properly updated.
+    pub fn handle_update(&mut self) {
```

Comment:
Let's call it `reset`?

Reply:
It's not true reset because it does not reset Pressed keys.